impl/src/main/java/org/pentaho/di/trans/dataservice/streaming/execution/StreamingServiceTransExecutor.java

```diff
@@ -0,0 +1,241 @@
+   */
+  private final Cache<String, StreamExecutionListener> serviceListeners = CacheBuilder.newBuilder()
+    .concurrencyLevel( 4 )
+    .expireAfterAccess( DataServiceConstants.STREAMING_CACHE_DURATION, DataServiceConstants.STREAMING_CACHE_TIME_UNIT )
+    .removalListener( new RemovalListener<String, StreamExecutionListener>() {
+      public void onRemoval( RemovalNotification<String, StreamExecutionListener> removal ) {
+        LogChannelInterface log = serviceTrans.getLogChannel();
+
+        removal.getValue().unSubscribe();
+
+        log.logDebug( DataServiceConstants.STREAMING_CACHE_REMOVED + removal.getKey() );
+      }
+    } )
+    .softValues()
+    .build();
+
+  private long lastCacheCleanupMillis = 0;
+
+  /**
+   * Constructor.
+   *
+   * @param id The instance id.
+   * @param serviceTrans The {@link org.pentaho.di.trans.Trans} Service Transformation.
+   * @param serviceStepName The Service Transformation Step Name.
+   */
+  public StreamingServiceTransExecutor( final String id, final Trans serviceTrans, final String serviceStepName ) {
+    this.serviceTrans = serviceTrans;
+    this.id = id;
+    this.serviceStepName = serviceStepName;
+  }
+
+  /**
+   * Getter for the {@link org.pentaho.di.trans.Trans} serviceTransformation
+   *
+   * @return {@link org.pentaho.di.trans.Trans} serviceTransformation
+   */
+  public Trans getServiceTrans() {
+    return serviceTrans;
+  }
+
+  /**
+   * Getter for the id.
+   *
+   * @return the id of this object instance.
+   */
+  public String getId() {
+    return id;
+  }
+
+  /**
+   * This method is used by the client to get the stream listener fot the given query and window parameters.
+   * If no cached listener exists it creates a new one, and spans the Service Transformation execution thread if not
+   * started, otherwise it returns the cached listener.
+   *
+   * @param query The requested query.
+   * @param windowSize The requested query window size.
+   * @param windowMillis The requested query rate.
+   * @return The {@link StreamExecutionListener} for the given query.
+   */
+  public StreamExecutionListener getBuffer( String query, int windowSize, long windowMillis ) {
+    String cacheId = getCacheKey( query, windowSize, windowMillis );
+
+    StreamExecutionListener streamListener = serviceListeners.getIfPresent( cacheId );
+
+    if ( streamListener == null ) {
+      if ( stepStream == null ) {
+        stepStream = new StreamList<>();
+      }
+
+      Observable<List<RowMetaAndData>> buffer = windowMillis > 0
+        ? windowSize > 0 ? stepStream.getStream().buffer( windowMillis, TimeUnit.MILLISECONDS, windowSize )
+        : stepStream.getStream().buffer( windowMillis, TimeUnit.MILLISECONDS )
+        : stepStream.getStream().buffer( windowSize );
+
+      streamListener = new StreamExecutionListener( buffer );
+
+      serviceListeners.put( cacheId, streamListener );
+    }
+
+    if ( !isRunning.get() ) {
+      isRunning.compareAndSet( false, true );
```

Comment:
You can just set, since the test was already done in the if statement. 
But if you really want atomic operations you should probably do the compareAndSet and then call startService() depending on the return value of the compareAndSet